File: ios/Goo/Bounds.swift
import Foundation
import Yaml

public class Bounds: Hashable, Equatable, CustomStringConvertible {
    let min: Vector
    let max: Vector
    
    public var hashValue: Int {
        return min.hashValue ^ max.hashValue &* 9973
    }
    
    public var description: String {
        return "Bounds(\(min.x), \(min.y), \(max.x), \(max.y))"
    }
    
    init(min: Vector, max: Vector) {
        self.min = min
        self.max = max
    }
    
    convenience init(minX: Double = 0, minY: Double = 0, maxX: Double = 0, maxY: Double = 0) {
        self.init(min: Vector(minX, minY), max: Vector(maxX, maxY))
    }
    
    convenience init?(_ yaml: Yaml) {
        if let min = Vector(yaml["min"]), let max = Vector(yaml["max"]) {
            self.init(min: min, max: max)
        } else {
            return nil
        }
    }
    
}

public func ==(left: Bounds, right: Bounds) -> Bool {
    return left.min == right.min && left.max == right.max
}

public func *(left: Bounds, right: Vector) -> Bounds {
    return Bounds(min: left.min * right, max: left.max * right)
}

File: android/app/src/main/java/org/haralovich/goo/Bounds.kt
package org.haralovich.goo

import com.fasterxml.jackson.databind.JsonNode

open class Bounds(val min: Vector, val max: Vector) {

    companion object {

        fun parse(json: JsonNode): Bounds? {
            val min = Vector.parse(json.path("min"))
            val max = Vector.parse(json.path("max"))
            if (min != null && max != null) {
                return Bounds(min, max)
            } else {
                return null
            }
        }

    }

    constructor(
        minX: Float = 0f,
        minY: Float = 0f,
        maxX: Float = 0f,
        maxY: Float = 0f
    ) : this (Vector(minX, minY), Vector(maxX, maxY))

    override fun equals(other: Any?): Boolean {
        if (other is Bounds) {
            return other.min == min && other.max == max
        } else {
            return false
        }
    }

    operator fun plus(inset: Inset): Bounds {
        return Bounds(
            min.x + inset.left,
            min.y + inset.top,
            max.x - inset.right,
            max.y - inset.bottom
        )
    }

    operator fun times(other: Vector): Bounds {
        return Bounds(min * other, max * other)
    }

    override fun toString(): String {
        return "Bounds(${min.x}, ${min.y}, ${max.x}, ${max.y})"
    }

}
